File: cpp/class_check.cpp
#include <iostream>
#include <unordered_map>
#include <string>
#include <chrono>

using namespace std;

// Single large class implementation
class LibrarySystem {
public:
    void addBook(int bookId, const string &title) {
        books[bookId] = title;
    }

    void addUser(int userId, const string &name) {
        users[userId] = name;
    }

    void borrowBook(int userId, int bookId) {
        if (books.count(bookId) && users.count(userId)) {
            if (!borrowedBooks.count(bookId)) {
                borrowedBooks[bookId] = userId;
            }
        }
    }

    void returnBook(int userId, int bookId) {
        if (borrowedBooks.count(bookId) && borrowedBooks[bookId] == userId) {
            borrowedBooks.erase(bookId);
        }
    }

private:
    unordered_map<int, string> books;
    unordered_map<int, string> users;
    unordered_map<int, int> borrowedBooks;
};

// Multiple smaller classes implementation
class Book {
public:
    Book(int bookId, const string &title) : bookId(bookId), title(title), isBorrowed(false) {}
    bool isBorrowed;
    int bookId;
    string title;
};

class User {
public:
    User(int userId, const string &name) : userId(userId), name(name) {}
    int userId;
    string name;
};

class Library {
public:
    void addBook(int bookId, const string &title) {
        books[bookId] = new Book(bookId, title);
    }

    void addUser(int userId, const string &name) {
        users[userId] = new User(userId, name);
    }

    void borrowBook(int userId, int bookId) {
        if (books.count(bookId) && users.count(userId)) {
            Book *book = books[bookId];
            if (!book->isBorrowed) {
                book->isBorrowed = true;
                borrowedBooks[bookId] = userId;
            }
        }
    }

    void returnBook(int userId, int bookId) {
        if (borrowedBooks.count(bookId) && borrowedBooks[bookId] == userId) {
            Book *book = books[bookId];
            book->isBorrowed = false;
            borrowedBooks.erase(bookId);
        }
    }

private:
    unordered_map<int, Book *> books;
    unordered_map<int, User *> users;
    unordered_map<int, int> borrowedBooks;
};

void measurePerformance(auto &library) {
    // Add a large number of books and users
    for (int bookId = 1; bookId <= 1000; ++bookId) {
        library.addBook(bookId, "Book " + to_string(bookId));
    }

    for (int userId = 1; userId <= 500; ++userId) {
        library.addUser(userId, "User " + to_string(userId));
    }

    // Perform borrow and return operations multiple times
    for (int i = 0; i < 10; ++i) {
        for (int userId = 1; userId <= 500; ++userId) {
            for (int bookId = 1; bookId <= 1000; ++bookId) {
                library.borrowBook(userId, bookId);
                library.returnBook(userId, bookId);
            }
        }
    }
}

int main() {
    // Measure performance of the single large class implementation
    LibrarySystem librarySystem;
    auto start = chrono::high_resolution_clock::now();
    measurePerformance(librarySystem);
    auto end = chrono::high_resolution_clock::now();
    cout << "Single large class implementation execution time: "
         << chrono::duration_cast<chrono::milliseconds>(end - start).count() << " ms" << endl;

    // Measure performance of the multiple smaller classes implementation
    Library library;
    start = chrono::high_resolution_clock::now();
    measurePerformance(library);
    end = chrono::high_resolution_clock::now();
    cout << "Multiple smaller classes implementation execution time: "
         << chrono::duration_cast<chrono::milliseconds>(end - start).count() << " ms" << endl;

    return 0;
}

File: py/tests/class_check.py
import time

# Single large class implementation
class LibrarySystem:
    def __init__(self):
        self.books = {}
        self.users = {}
        self.borrowed_books = {}

    def add_book(self, book_id, title):
        self.books[book_id] = title

    def add_user(self, user_id, name):
        self.users[user_id] = name

    def borrow_book(self, user_id, book_id):
        if book_id in self.books and user_id in self.users:
            if book_id not in self.borrowed_books:
                self.borrowed_books[book_id] = user_id

    def return_book(self, user_id, book_id):
        if book_id in self.borrowed_books and self.borrowed_books[book_id] == user_id:
            del self.borrowed_books[book_id]

    def show_books(self):
        pass

    def show_users(self):
        pass

# Multiple smaller classes implementation
class Book:
    def __init__(self, book_id, title):
        self.book_id = book_id
        self.title = title
        self.is_borrowed = False

class User:
    def __init__(self, user_id, name):
        self.user_id = user_id
        self.name = name

class Library:
    def __init__(self):
        self.books = {}
        self.users = {}
        self.borrowed_books = {}

    def add_book(self, book_id, title):
        self.books[book_id] = Book(book_id, title)

    def add_user(self, user_id, name):
        self.users[user_id] = User(user_id, name)

    def borrow_book(self, user_id, book_id):
        if book_id in self.books and user_id in self.users:
            book = self.books[book_id]
            if not book.is_borrowed:
                book.is_borrowed = True
                self.borrowed_books[book_id] = user_id

    def return_book(self, user_id, book_id):
        if book_id in self.borrowed_books and self.borrowed_books[book_id] == user_id:
            book = self.books[book_id]
            book.is_borrowed = False
            del self.borrowed_books[book_id]

    def show_books(self):
        pass

    def show_users(self):
        pass

def measure_performance(library):
    # Add a large number of books and users
    for book_id in range(1, 1001):
        library.add_book(book_id, f"Book {book_id}")
    
    for user_id in range(1, 501):
        library.add_user(user_id, f"User {user_id}")

    # Perform borrow and return operations multiple times
    for _ in range(10):
        for user_id in range(1, 501):
            for book_id in range(1, 1001):
                library.borrow_book(user_id, book_id)
                library.return_book(user_id, book_id)

if __name__ == "__main__":
    # Measure performance of the single large class implementation
    library_system = LibrarySystem()
    start_time = time.time()
    measure_performance(library_system)
    print(f"Single large class implementation execution time: {time.time() - start_time:.6f} seconds")

    # Measure performance of the multiple smaller classes implementation
    library = Library()
    start_time = time.time()
    measure_performance(library)
    print(f"Multiple smaller classes implementation execution time: {time.time() - start_time:.6f} seconds")